libs/vkd3d-shader/spirv.c

```diff
@@ -8385,6 +8735,52 @@ static uint32_t vkd3d_dxbc_compiler_get_resource_index(struct vkd3d_dxbc_compile
             vkd3d_dxbc_compiler_load_descriptor_table_offset(compiler, descriptor_table),
             index_id);
 
+    /* Inject a check for descriptor type here. */
+    if (compiler->descriptor_qa_check_func_id && binding->type != VKD3D_SHADER_DESCRIPTOR_TYPE_SAMPLER)
+    {
+        switch (binding->type)
+        {
+            case VKD3D_SHADER_DESCRIPTOR_TYPE_CBV:
+                if (compiler->shader_interface.flags & VKD3D_SHADER_INTERFACE_BINDLESS_CBV_AS_STORAGE_BUFFER)
+                    type_flags = VKD3D_DESCRIPTOR_QA_TYPE_STORAGE_BUFFER_BIT;
+                else
+                    type_flags = VKD3D_DESCRIPTOR_QA_TYPE_UNIFORM_BUFFER_BIT;
+                break;
+
+            case VKD3D_SHADER_DESCRIPTOR_TYPE_UAV:
+                if (binding->flags & VKD3D_SHADER_BINDING_FLAG_IMAGE)
+                    type_flags = VKD3D_DESCRIPTOR_QA_TYPE_STORAGE_IMAGE_BIT;
+                else if (binding->flags & VKD3D_SHADER_BINDING_FLAG_RAW_SSBO)
+                    type_flags = VKD3D_DESCRIPTOR_QA_TYPE_STORAGE_BUFFER_BIT;
+                else
+                    type_flags = VKD3D_DESCRIPTOR_QA_TYPE_STORAGE_TEXEL_BUFFER_BIT;
+                break;
+
+            case VKD3D_SHADER_DESCRIPTOR_TYPE_SRV:
+                if (binding->flags & VKD3D_SHADER_BINDING_FLAG_IMAGE)
+                    type_flags = VKD3D_DESCRIPTOR_QA_TYPE_SAMPLED_IMAGE_BIT;
+                else if (binding->flags & VKD3D_SHADER_BINDING_FLAG_RAW_SSBO)
+                    type_flags = VKD3D_DESCRIPTOR_QA_TYPE_STORAGE_BUFFER_BIT;
+                else
+                    type_flags = VKD3D_DESCRIPTOR_QA_TYPE_UNIFORM_TEXEL_BUFFER_BIT;
+                break;
+
+            default:
+                ERR("Invalid binding type %u.\n", binding->type);
+                type_flags = 0;
+                break;
+        }
+
+        descriptor_qa_args[0] = index_id;
+        descriptor_qa_args[1] = vkd3d_dxbc_compiler_get_constant_uint(compiler, type_flags);
+        descriptor_qa_args[2] = vkd3d_dxbc_compiler_get_constant_uint(compiler,
+                ++compiler->descriptor_qa_instruction_count);
+        index_id = vkd3d_spirv_build_op_function_call(builder,
```

Comment:
is this valid? spirv-val already complains about our hull shader phase functions taking raw IDs rather than variables.

Reply:
It is. It's when passing pointers that the pointer must bind directly to an OpVariable declaration.

Reply:
Verified to be sure, and it passes spirv-val just fine.